### PR TITLE
fix memtest and print out total free memory/page

### LINKED ISSUE #21: Unused memtest code could be replaced by e820
https://github.com/crazyss/hypervisor_last/blob/8a6b11f9a9dc9ba7d0cdb1496abf1847c8744216/hypervisor/start.c#L133
This code will cause the system crash due to paging fault, (we don't have page fault handler right now).
So, originally, I use this code to compute how many memory do we have. for now we should use e820.
@coolgw would you please take a look.

remove unused code, read the real data from e820 table, and display them on the screen.

### PR DESCRIPTION
**Pull request by coolgw**
PR content include:
1) Base result of memory detection of e820, we start do memtest on each free area.
2) Put usable info into the mem_map

### CODE REVIEW

File: hypervisor/Makefile [hunk 1]
```diff
@@ -1,7 +1,7 @@
 
 
 ASFLAGS = --32 -march=i486 -mtune=i486
-CFLAGS = -m32 -I../include
+CFLAGS = -m32 -fno-stack-protector -I../include
```
Comment (wjn740) [inline hypervisor/Makefile:4]: why add this ?

Comment (coolgw) [reply hypervisor/Makefile:4]: This will fix an issue for ubuntu env. 
Add this can let your code compile pass on ubuntu.

**Comment by wjn740** [comment]
Would you mind giving some structures of design?
So that we could understand easier? for now, the description is empty, might be you could add something.

**Comment by coolgw** [comment]
> 
> 
> Would you mind giving some structures of design?
> So that we could understand easier? for now, the description is empty, might be you could add something.

Updated, you can check again.

File: hypervisor/start.c [hunk 2]
```diff
@@ -131,7 +135,6 @@ static void drawing_desktop()
     boxfill8(vram, xsize, COL8_FFFFFF, xsize - 47, ysize -  3, xsize -  4, ysize -  3);
     boxfill8(vram, xsize, COL8_FFFFFF, xsize -  3, ysize - 24, xsize -  3, ysize -  3);
 
-    memtotal=memtest(0x400000, 0xbfffffff);
     ret = memman_free((struct MEMMAN*)(MEMMAN_ADDR), 0x00100000, 0x00100000);
```
Comment (wjn740) [inline hypervisor/start.c]: this memman_free also is old code. I think it could be removed or override.

File: include/common.h [hunk 3]
```diff
@@ -44,6 +44,17 @@ unsigned int memtest(unsigned int start, unsigned int end);
 #define COL8_008484     14
 #define COL8_848484     15
 #define MEM_MAP_ADDR (char *)((0x00078004)-(SYSSEG << 4))
+#define MAX_FREE_MEM_BLOCK 10
+#define LOW_MEM 0x10000 //current igd base address                        
+#define PAGING_MEMORY (64 * 1024 * 1024)  //currently we map 64M , i suppose currently is enough        
+#define PAGING_PAGES (PAGING_MEMORY >> 12)       
+#define MAP_NR(addr) (((addr)) >> 12)
```
Comment (wjn740) [inline include/common.h]: Is it Page_frame number?

File: include/common.h [hunk 4]
```diff
@@ -44,6 +44,17 @@ unsigned int memtest(unsigned int start, unsigned int end);
 #define COL8_008484     14
 #define COL8_848484     15
 #define MEM_MAP_ADDR (char *)((0x00078004)-(SYSSEG << 4))
+#define MAX_FREE_MEM_BLOCK 10
+#define LOW_MEM 0x10000 //current igd base address                        
+#define PAGING_MEMORY (64 * 1024 * 1024)  //currently we map 64M , i suppose currently is enough
```
Comment (wjn740) [inline include/common.h]: I think PAGING_MEMORY should be detected via e820 area.

File: include/common.h [hunk 5]
```diff
@@ -44,6 +44,17 @@ unsigned int memtest(unsigned int start, unsigned int end);
 #define COL8_008484     14
 #define COL8_848484     15
 #define MEM_MAP_ADDR (char *)((0x00078004)-(SYSSEG << 4))
+#define MAX_FREE_MEM_BLOCK 10
+#define LOW_MEM 0x10000 //current igd base address
```
Comment (wjn740) [inline include/common.h]: What is IGD, I dump register info at your codebase. 
CS register is 0x10, it point to 0x10000.

```
(qemu) info  registers 
EAX=00000000 EBX=00000000 ECX=000ffebe EDX=0000000f
ESI=00000108 EDI=000080a7 EBP=000fffec ESP=000fff10
EIP=00001b71 EFL=00000206 [-----P-] CPL=0 II=0 A20=1 SMM=0 HLT=1
ES =0018 00010000 ffffffff 00cf9300 DPL=0 DS   [-WA]
CS =0010 00010000 ffffffff 00cf9a00 DPL=0 CS32 [-R-]
SS =0018 00010000 ffffffff 00cf9300 DPL=0 DS   [-WA]
DS =0018 00010000 ffffffff 00cf9300 DPL=0 DS   [-WA]
FS =0018 00010000 ffffffff 00cf9300 DPL=0 DS   [-WA]
GS =0018 00010000 ffffffff 00cf9300 DPL=0 DS   [-WA]
LDT=0000 00000000 0000ffff 00008200 DPL=0 LDT
TR =0000 00000000 0000ffff 00008b00 DPL=0 TSS32-busy
GDT=     000702fa 00000020
IDT=     0026f800 000007ff
CR0=00000011 CR2=00000000 CR3=00021000 CR4=00000000
DR0=0000000000000000 DR1=0000000000000000 DR2=0000000000000000 DR3=0000000000000000 
DR6=00000000ffff0ff0 DR7=0000000000000400
EFER=0000000000000000
FCW=037f FSW=0000 [ST=0] FTW=00 MXCSR=00001f80
FPR0=0000000000000000 0000 FPR1=0000000000000000 0000
FPR2=0000000000000000 0000 FPR3=0000000000000000 0000
FPR4=0000000000000000 0000 FPR5=0000000000000000 0000
FPR6=0000000000000000 0000 FPR7=0000000000000000 0000
XMM00=00000000000000000000000000000000 XMM01=00000000000000000000000000000000
XMM02=00000000000000000000000000000000 XMM03=00000000000000000000000000000000
XMM04=00000000000000000000000000000000 XMM05=00000000000000000000000000000000
XMM06=00000000000000000000000000000000 XMM07=00000000000000000000000000000000
```

File: hypervisor/start.c [hunk 6]
```diff
@@ -163,15 +184,28 @@ void drawing_mem_map()
 		putfont8_string(vram, xsize, 150, pos, COL8_FFFFFF, font.Bitmap,
 				buf);
 		if (memmap->type == 1) {
+            free_memory[i][0] = memmap->base;
+            free_memory[i][1] = memmap->length;
+            memtest(memmap->base, memmap->base + memmap->length - 1);
```
Comment (wjn740) [inline hypervisor/start.c]: memtest(memmap->base, memmap->length - 1);  looks better.

compute 'memmap->base + memmap->length -1' in function.

File: include/common.h (hunk 4 above)
Comment (coolgw) [reply include/common.h]: Add max_paging_mem , this is store the max paging value base e820.

File: hypervisor/start.c (hunk 6 above)
Comment (coolgw) [reply hypervisor/start.c]: the function need put the end address, not the length.
unsigned int memtest(unsigned int start, unsigned int end);

File: include/common.h (hunk 3 above)
Comment (coolgw) [reply include/common.h]: right

File: include/common.h (hunk 5 above)
Comment (coolgw) [reply include/common.h]: gdt base address, this not used now, i can remove it.

File: hypervisor/start.c (hunk 6 above)
Comment (coolgw) [reply hypervisor/start.c]: will update the memtest api interface.

File: include/common.h (hunk 4 above)
Comment (coolgw) [reply include/common.h]: as agreement, we increase the size to 512M

**Comment by wjn740** [comment]
I tested the latest version. It is OK. I will merge.

**Pull request merged**